### ISSUE TITLE
debugger: o warn=0 die=0 ignored

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9421]
Migrated from [rt.perl.org#57016](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=57016) (status was 'resolved')

Searchable as RT57016$

**Comment by p5pRT** [comment]
### From tye.mcqueen@gmail.com

#### Created by tyemq@cpan.org



In the Perl debugger\, the command "o warn=0 die=0" does nothing useful\.

This bug also applies to the latest 5\.010 Perl I could conveniently find
on http&#8203;://search\.cpan\.org/\.

Here is a fix&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- 5.8.7/perl5db.pl    2005-09-27 10:45:51 -0700
+++ tye/perl5db.pl    2008-07-16 12:52:18 -0700
@@ -7703,6 +7703,8 @@
         }
         elsif ($prevwarn) {
             $SIG{__WARN__} = $prevwarn;
+        } else {
+            undef $SIG{__WARN__};
         }
     } ## end if (@_)
     $warnLevel;
@@ -7744,6 +7746,9 @@
         elsif ($prevdie) {
             $SIG{__DIE__} = $prevdie;
             print $OUT "Default die handler restored.\n";
+        } else {
+            undef $SIG{__DIE__};
+            print $OUT "Die handler removed.\n";
         }
     } ## end if (@_)
     $dieLevel;
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.7:

Configured by bcundal at Mon Nov 14 11:08:07 PST 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=linux, osvers=2.6.10w3data-web-1, archname=i686-linux
    uname='linux dasdev1 2.6.10w3data-web-1 #4 smp wed feb 2 09:46:25 pst
2005 i686 i686 i386 gnulinux '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.2.3 20030502 (Red Hat Linux 3.2.3-42)',
gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.7:
    /home/tmcqueen/cvs/da/common/lib
    /home/tmcqueen/cvs/da/das2/lib
    /usr/local/lib/perl5/5.8.7/i686-linux
    /usr/local/lib/perl5/5.8.7
    /usr/local/lib/perl5/site_perl/5.8.7/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.7
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.7:
    HOME=/home/tmcqueen
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/opt/dataflux/client/lib/:
    LOGDIR (unset)

PATH=/usr/local/bin:/usr/local/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/opt/oracle/product/current/bin:/usr/local/dev-util/bin:/opt/oracle/product/current/bin:/home/tmcqueen/bin
    PERL5LIB=/home/tmcqueen/cvs/da/common/lib:/home/tmcqueen/cvs/da/das2/lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Wed Jul 16 13&#8203;:02&#8203;:26 2008\, tye\.mcqueen@&#8203;gmail\.com wrote&#8203;:
> This is a bug report for perl from tyemq@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> In the Perl debugger\, the command "o warn=0 die=0" does nothing
> useful\.
> 
> This bug also applies to the latest 5\.010 Perl I could conveniently
> find
> on http&#8203;://search\.cpan\.org/\.
> 
> Here is a fix&#8203;:
> 
> \-\-\- 5\.8\.7/perl5db\.pl    2005\-09\-27 10&#8203;:45&#8203;:51 \-0700
> \+\+\+ tye/perl5db\.pl    2008\-07\-16 12&#8203;:52&#8203;:18 \-0700
> @&#8203;@&#8203; \-7703\,6 \+7703\,8 @&#8203;@&#8203;
>          \}
>          elsif \($prevwarn\) \{
>              $SIG\{\_\_WARN\_\_\} = $prevwarn;
> \+        \} else \{
> \+            undef $SIG\{\_\_WARN\_\_\};
>          \}
>      \} \#\# end if \(@&#8203;\_\)
>      $warnLevel;
> @&#8203;@&#8203; \-7744\,6 \+7746\,9 @&#8203;@&#8203;
>          elsif \($prevdie\) \{
>              $SIG\{\_\_DIE\_\_\} = $prevdie;
>              print $OUT "Default die handler restored\.\\n";
> \+        \} else \{
> \+            undef $SIG\{\_\_DIE\_\_\};
> \+            print $OUT "Die handler removed\.\\n";
>          \}
>      \} \#\# end if \(@&#8203;\_\)
>      $dieLevel;
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-


Thanks for the bug report/patch\.

Can you also provide an example of where it doesn't behave as you 
expected?


Kind regards\,

Bram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From tye.mcqueen@gmail.com

On Sat\, Jul 26\, 2008 at 11&#8203;:30 AM\, Bram via RT \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> Thanks for the bug report/patch\.
>
> Can you also provide an example of where it doesn't behave as you
> expected?
>

Um\, just start the debugger\, type "o warn=0 die=0" and then use warn or die
and notice that lots of extraneous stack trace is still displayed\.

If you are unable to notice the difference\, then just simulate what it
should do via&#8203;:

$SIG\{\_\_WARN\_\_\}= $SIG\{\_\_DIE\_\_\}= 0

and then try warn/die and see how unadorned the output is\.

Oh\, to start the debugger\, you can use "perl \-debug"\.  A full demonstration&#8203;:

========================================

 $ perl \-debug

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   bug
&nbsp;  DB\<1> warn "extra junk\!\\n"
extra junk\!
 at \(eval 5\)\[/usr/local/lib/perl5/5\.8\.7/perl5db\.pl&#8203;:628\] line 2
&nbsp;        eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D =
$^D |
 $DB&#8203;::db\_stop;
warn "extra junk\!\\\\n";

;' called at /usr/local/lib/perl5/5\.8\.7/perl5db\.pl line 628
&nbsp;        DB&#8203;::eval called at /usr/local/lib/perl5/5\.8\.7/perl5db\.pl line 3410
&nbsp;        DB&#8203;::DB called at \-e line 1

&nbsp;  DB\<2> o warn=0
&nbsp;           warnLevel = '0'
&nbsp;  DB\<3> warn "junk?\\n"
junk?
 at \(eval 6\)\[/usr/local/lib/perl5/5\.8\.7/perl5db\.pl&#8203;:628\] line 2
&nbsp;        eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D =
$^D |
 $DB&#8203;::db\_stop;
warn "junk?\\\\n";

;' called at /usr/local/lib/perl5/5\.8\.7/perl5db\.pl line 628
&nbsp;        DB&#8203;::eval called at /usr/local/lib/perl5/5\.8\.7/perl5db\.pl line 3410
&nbsp;        DB&#8203;::DB called at \-e line 1

&nbsp;  DB\<4> $SIG\{\_\_WARN\_\_\}= 0

&nbsp;  DB\<5> warn "just this\!\\n"
just this\!

&nbsp;  DB\<6> q
 $

========================================

Note that I've verified that this is broken in standard 5\.8\.7\, 5\.8\.8\, and
5\.10\.  However\, Strawberry Perl 5\.8\.8 appears to have patched this to fix it
and also to set the default warnLevel to 0 instead of to 1\, as shown below\.
Does it really take this long for patches to arrive from Strawberry or just
for them to get applied?

========================================

 > perl \-debug

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`perldoc perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   bug

SetConsoleMode failed\, LastError=|6| at
C&#8203;:/strawberry\-perl/perl/site/lib/Term/Re
adKey\.pm line 265\.
&nbsp;  DB\<1> o warn?
&nbsp;           warnLevel = '0'

&nbsp;  DB\<2> warn "just this\\n"
just this


&nbsp;  DB\<3> o warn=1
&nbsp;           warnLevel = '1'

&nbsp;  DB\<4> warn "junk\!\\n"
junk\!
 at \(eval 11\)\[C&#8203;:/strawberry\-perl/perl/lib/perl5db\.pl&#8203;:628\] line 2
&nbsp;        eval '\($@&#8203;\, $\!\, $^E\, $\,\, $/\, $\\\\\, $^W\) = @&#8203;saved;package main; $^D =
$^D |
 $DB&#8203;::db\_stop;
warn "junk\!\\\\n";

;' called at C&#8203;:/strawberry\-perl/perl/lib/perl5db\.pl line 628
&nbsp;        DB&#8203;::eval called at C&#8203;:/strawberry\-perl/perl/lib/perl5db\.pl line 3410
&nbsp;        DB&#8203;::DB called at \-e line 1


&nbsp;  DB\<5> o warn=0
&nbsp;           warnLevel = '0'

&nbsp;  DB\<6> warn "Just this\!\\n"
Just this\!


&nbsp;  DB\<7> q

 >

========================================

Oh\, the Strawberry Perl fix is simpler than the fix I proposed\, just replace
the "elsif"s with plain "else"s&#8203;:

sub warnLevel \{
&nbsp;    if \(@&#8203;\_\) \{
&nbsp;        $prevwarn = $SIG\{\_\_WARN\_\_\} unless $warnLevel;
&nbsp;        $warnLevel = shift;
&nbsp;        if \($warnLevel\) \{
&nbsp;            $SIG\{\_\_WARN\_\_\} = \\&DB&#8203;::dbwarn;
&nbsp;        \}
&nbsp;        else \{
&nbsp;            $SIG\{\_\_WARN\_\_\} = $prevwarn;
&nbsp;        \}
&nbsp;    \} \#\# end if \(@&#8203;\_\)
&nbsp;    $warnLevel;
\} \#\# end sub warnLevel

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Citeren Tye McQueen \<tye\.mcqueen@&#8203;gmail\.com>&#8203;:

> On Sat\, Jul 26\, 2008 at 11&#8203;:30 AM\, Bram via RT   
> \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:
>
>> Thanks for the bug report/patch\.
>>
>> Can you also provide an example of where it doesn't behave as you
>> expected?
>>
>
> Um\, just start the debugger\, type "o warn=0 die=0" and then use warn or die
> and notice that lots of extraneous stack trace is still displayed\.
>
> If you are unable to notice the difference\, then just simulate what it
> should do via&#8203;:
>
> $SIG\{\_\_WARN\_\_\}= $SIG\{\_\_DIE\_\_\}= 0
>
> and then try warn/die and see how unadorned the output is\.

I see\.

A small note tho&#8203;: o warnLevel 0  is not intended to simulate  
$SIG\{\_\_WARN\_\_\} = 0\.

What it is supposed to do is restore the previously set warning handler\.

Currently if no previously handler is set then  o warnLevel 0  is a NUL op\.

I also feel that this is not very useful \(but I don't use \-d very  
often\) and that your patch should be applied\.


>
> Note that I've verified that this is broken in standard 5\.8\.7\, 5\.8\.8\, and
> 5\.10\.  However\, Strawberry Perl 5\.8\.8 appears to have patched this to fix it
> and also to set the default warnLevel to 0 instead of to 1\, as shown below\.
> Does it really take this long for patches to arrive from Strawberry or just
> for them to get applied?

Note that you can set the default options via an rc file\. \(look at  
perldoc perl5db for more information\)\.

> Oh\, the Strawberry Perl fix is simpler than the fix I proposed\, just replace
> the "elsif"s with plain "else"s&#8203;:
>
> sub warnLevel \{
>     if \(@&#8203;\_\) \{
>         $prevwarn = $SIG\{\_\_WARN\_\_\} unless $warnLevel;
>         $warnLevel = shift;
>         if \($warnLevel\) \{
>             $SIG\{\_\_WARN\_\_\} = \\&DB&#8203;::dbwarn;
>         \}
>         else \{
>             $SIG\{\_\_WARN\_\_\} = $prevwarn;
>         \}
>     \} \#\# end if \(@&#8203;\_\)
>     $warnLevel;
> \} \#\# end sub warnLevel

May I ask at what Strawberry Perl you are looking?

I just downloaded Strawberry Perl 5\.10\.0\.1 and Strawberry Perl 5\.8\.8\.1  
and both have&#8203;:

sub warnLevel \{
&nbsp;     if \(@&#8203;\_\) \{
&nbsp;         $prevwarn = $SIG\{\_\_WARN\_\_\} unless $warnLevel;
&nbsp;         $warnLevel = shift;
&nbsp;         if \($warnLevel\) \{
&nbsp;             $SIG\{\_\_WARN\_\_\} = \\&DB&#8203;::dbwarn;
&nbsp;         \}
&nbsp;         elsif \($prevwarn\) \{
&nbsp;             $SIG\{\_\_WARN\_\_\} = $prevwarn;
&nbsp;         \}
&nbsp;     \} \#\# end if \(@&#8203;\_\)
&nbsp;     $warnLevel;
\} \#\# end sub warnLevel



Kind regards\,

Bram

**Comment by p5pRT** [comment]
### From tye.mcqueen@gmail.com

On Sat\, Aug 2\, 2008 at 10&#8203;:45 AM\, Bram \<p5p@&#8203;perl\.wizbit\.be> wrote&#8203;:

> A small note tho&#8203;: o warnLevel 0  is not intended to simulate $SIG\{\_\_WARN\_\_\}
> = 0\.
>
> What it is supposed to do is restore the previously set warning handler\.
>
> Currently if no previously handler is set then  o warnLevel 0  is a NUL op\.
>

No\, it is supposed to restore the previous state of the warn handle\.  Having
no warn handler is a previous state\.  It worked properly before \(prior to
the refactoring that changed "O" to "o"\)\.

May I ask at what Strawberry Perl you are looking?
>

Maybe I patched it myself long ago and forgot about it\.

Tye

**Comment by p5pRT** [comment]
### From @rgs

2008/7/16 via RT Tye McQueen \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> In the Perl debugger\, the command "o warn=0 die=0" does nothing useful\.
>
> This bug also applies to the latest 5\.010 Perl I could conveniently find
> on http&#8203;://search\.cpan\.org/\.
>
> Here is a fix&#8203;:
>
> \-\-\- 5\.8\.7/perl5db\.pl    2005\-09\-27 10&#8203;:45&#8203;:51 \-0700
> \+\+\+ tye/perl5db\.pl    2008\-07\-16 12&#8203;:52&#8203;:18 \-0700

Thanks\, applied as change \#34312\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
